lib/astutils.cpp

```diff
@@ -3647,3 +3647,82 @@ bool isUnevaluated(const Token *tok)
 {
     return Token::Match(tok, "alignof|_Alignof|_alignof|__alignof|__alignof__|decltype|offsetof|sizeof|typeid|typeof|__typeof__ (");
 }
+
+static std::set<std::string> getSwitchExprs(const Token *startbrace, bool &hasDefault)
+{
+    std::set<std::string> exprs;
+    const Token *endbrace = startbrace->link();
+    if (!endbrace)
+        return {};
+
+    hasDefault = false;
+    for (const Token *tok = startbrace->next(); tok && tok != endbrace; tok = tok->next()) {
+        if (Token::simpleMatch(tok, "{")) {
+            tok = tok->link();
+            continue;
+        }
+        if (Token::simpleMatch(tok, "case")) {
+            const Token *exprtok = tok->astOperand1();
+            if (!exprtok)
+                continue;
+            while (Token::Match(tok, "%name% ::"))
+                tok = tok->tokAt(2);
+            exprs.insert(exprtok->str());
+            continue;
+        }
+        if (Token::simpleMatch(tok, "default")) {
+            hasDefault = true;
+            continue;
+        }
+    }
+
+    return exprs;
+}
+
+bool isExhaustiveSwitch(const Token *startbrace)
+{
+    if (!startbrace)
+        return false;
+    const Token *rpar = startbrace->previous();
+    if (!rpar)
+        return false;
+    const Token *lpar = rpar->link();
+    if (!lpar)
+        return false;
+    const Token *vartok = lpar->next();
```

Comment:
you don't check that start brace is a `{`, what rpar and lpar is. they could be random other tokens.

more strict checking:
```
     if (!startbrace || !Token::simpleMatch(startbrace->previous(), ") {") || startbrace->scope()->type != Scope::ScopeType::eSwitch)
        return false;
```